src/plugins/message-parser/MessageParser.js

```diff
@@ -0,0 +1,86 @@
+import { html } from 'lit-html';
+import { pipe } from '@converse/headless/utils/functional';
+import RichMessage from './RichMessage';
+import log from '@converse/headless/log';
+
+class MessageParser {
+
+    constructor () {
+        this._filters = [];
+        this._middleware = [];
+    }
+
+    get filters () {
+        return this._filters;
+    }
+
+    get middleware () {
+        return this._middleware;
+    }
+
+    static checkReferenceCollision (reference, rich_message) {
+        const collides = rich_message.references.some(msg_ref => {
+          const is_after = reference.begin > msg_ref.end;
+          const is_before = reference.end < msg_ref.begin;
+          return !is_after && !is_before;
+        });
+        return collides;
+    }
+
+    static makeTemplateResultFromString (string_or_templateresult) {
+        const is_template_result = !!string_or_templateresult.getTemplateElement;
+        return is_template_result
+          ? string_or_templateresult
+          : html`${string_or_templateresult}`;
+    }
+
+    static addSingleReferenceToMessage (reference, rich_message) {
+        if (!MessageParser.checkReferenceCollision(reference, rich_message)) {
+            const { begin, end, template } = reference;
+            const template_result = MessageParser.makeTemplateResultFromString(template);
+            rich_message.addTemplateResult(begin, end, template_result);
+        } else {
+            throw new Error('Reference collision');
+        }
+    }
+
+    static addReferencesToMessage (references, rich_message) {
+        references.forEach(reference =>
+            MessageParser.addSingleReferenceToMessage(reference, rich_message)
+        );
+    }
+
+    addFilters (...filters) {
+        this._filters.push(...filters);
+    }
+
+    addMiddleware (...middleware) {
+        this._middleware.push(...middleware);
+    }
```

Comment:
I like the idea of having a middleware pipeline that 3rd party plugins can use to add more rich content to messages.

The middleware will have to conform to some strict rules. For example, we can't have rich parts of the message overlap one another.